codecs/imagequant/imagequant.c

```diff
@@ -50,6 +52,169 @@ void quantize(uint8_t* image_buffer, int image_width, int image_height, int num_
+ * both be 'regular' or 'bright'. Black exists as both regular and bright.
+ */
+EMSCRIPTEN_KEEPALIVE
+void zx_quantize(uint8_t* image_buffer, int image_width, int image_height, float dithering) {
+  int size = image_width * image_height;
+  int bytes_per_pixel = 4;
+  result = (int) malloc(size * bytes_per_pixel);
+  uint8_t* block = malloc(8 * 8 * bytes_per_pixel);
+  uint8_t* image8bit = (uint8_t*) malloc(8 * 8);
+
+  // For each 8x8 grid
+  for (int block_start_y = 0; block_start_y < image_height; block_start_y += 8) {
+    for (int block_start_x = 0; block_start_x < image_width; block_start_x += 8) {
+      int color_popularity[15] = {0};
+      int block_index = 0;
+      int block_width = 8;
+      int block_height = 8;
+
+      // If the block hangs off the right/bottom of the image dimensions, make it smaller to fit.
+      if (block_start_y + block_height > image_height) {
+        block_height = image_height - block_start_y;
+      }
+
+      if (block_start_x + block_width > image_width) {
+        block_width = image_width - block_start_x;
+      }
+
+      // For each pixel in that block:
+      for (int y = block_start_y; y < block_start_y + block_height; y++) {
+        for (int x = block_start_x; x < block_start_x + block_width; x++) {
+          int pixel_start = (y * image_width * bytes_per_pixel) + (x * bytes_per_pixel);
+          int smallest_distance = INT_MAX;
+          int winning_index = -1;
+
+          // Copy pixel data for quantizing later
+          block[block_index++] = image_buffer[pixel_start];
+          block[block_index++] = image_buffer[pixel_start + 1];
+          block[block_index++] = image_buffer[pixel_start + 2];
+          block[block_index++] = image_buffer[pixel_start + 3];
+
+          // Which zx color is this pixel closest to?
+          for (int color_index = 0; color_index < 15; color_index++) {
+            liq_color color = zx_colors[color_index];
+
+            // Using Euclidean distance. LibQuant has better methods, but it requires conversion to
+            // LAB, so I don't think it's worth it.
+            int distance =
+              pow(color.r - image_buffer[pixel_start + 0], 2) +
+              pow(color.g - image_buffer[pixel_start + 1], 2) +
+              pow(color.b - image_buffer[pixel_start + 2], 2);
+
+            if (distance < smallest_distance) {
+              winning_index = color_index;
+              smallest_distance = distance;
+            }
+          }
+          color_popularity[winning_index]++;
+        }
+      }
+
+      // Get the three most popular colours for the block.
+      int first_color_index = 0;
+      int second_color_index = 0;
+      int third_color_index = 0;
+      int highest_popularity = -1;
+      int second_highest_popularity = -1;
+      int third_highest_popularity = -1;
+
+      for (int color_index = 0; color_index < 15; color_index++) {
```

Comment:
I think there’s a more elegant way to do this block, but also it’s an easter egg. I might grab this in my spare time if I fee like it 😅

Reply:
Interested to see the better way of doing this. My unfamiliarity with C was getting in the way.